### PR TITLE
Implemented support for player location

### PR DESCRIPTION
**Pull request by Podshot**
Adds a player manager that handles retrieving and provided Player objects, which contain the dimension, location and rotation of a given player.
Added load error and does not exist error for players.

Started reworking dimension managing. Made dimensions namespaced.
Added contains and iter method to the LevelDB class.
Added tests for the leveldb class.
Moved more logic into DatabaseHistoryManager from the chunk manager.
Added more example worlds.

### CODE REVIEW

File: amulet/level/formats/leveldb_world/format.py [hunk 1]
```diff
@@ -310,3 +311,37 @@ def _get_raw_chunk_data(
...
+        """
+        if player_id == LOCAL_PLAYER:
+            key = player_id
+        else:
+            key = f"player_{player_id}".encode("utf-8")
+        try:
+            data = self._level_manager._db.get(key)
+        except KeyError:
+            raise KeyError(f"Player {player_id} doesn't exist")
+        player_nbt = nbt.load(buffer=data, compressed=False, little_endian=True)
```
Comment (gentlegiantJGC) [inline amulet/level/formats/leveldb_world/format.py:341]: The buffer key word argument has been removed from the nbt load function.
The first input can be either a string path to a file to load, a file-like object (with a read method) or a bytes object of the actual nbt data. It should look like this
```
nbt.load(data, compressed=False, little_endian=True)
```

File: amulet/api/player/player_manager.py [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+from amulet.api.history.history_manager import DatabaseHistoryManager
+from amulet.api import level as api_level
+
+LOCAL_PLAYER = "~local_player"
+
+
+class Player(Changeable):
+    def __init__(
+        self,
+        _uuid: str,
```
Comment (gentlegiantJGC) [inline amulet/api/player/player_manager.py:18]: This shouldn't have the leading underscore

File: amulet/level/formats/anvil_world/format.py [hunk 3]
```diff
@@ -391,6 +392,48 @@ def _get_raw_chunk_data(
         """
         return self._get_dimension(dimension).get_chunk_data(cx, cz)
 
+    def get_players(self) -> Generator[str, None, None]:
```
Comment (gentlegiantJGC) [inline amulet/level/formats/anvil_world/format.py:395]: I would prefer if this was `all_player_ids` so that it isn't confused with `get_players` and to match the `all_chunk_coords` method.

File: amulet/level/formats/anvil_world/format.py [hunk 4]
```diff
@@ -391,6 +392,48 @@ def _get_raw_chunk_data(
...
+        """
+        Returns a generator of all player ids that are present in the level
+        """
+        yield from (
+            os.path.splitext(pid)[0]
+            for pid in (
+                os.path.basename(f)
+                for f in glob.iglob(os.path.join(self.path, "playerdata", "*.dat"))
+            )
+        )
```
Comment (gentlegiantJGC) [inline amulet/level/formats/anvil_world/format.py:405]: This is a more compact way of writing this.
```
for f in glob.iglob(os.path.join(self.path, "playerdata", "*.dat")):
    yield os.path.splitext(os.path.basename(f))[0]
```

File: amulet/level/formats/anvil_world/format.py [hunk 5]
```diff
@@ -391,6 +392,48 @@ def _get_raw_chunk_data(
...
+
+        path = os.path.join(self.path, "playerdata", f"{player_id}.dat")
+        if not os.path.exists(path):
+            raise KeyError(f"Player {player_id} doesn't exist")
+        player_nbt = nbt.load(path)
+        return Player(
+            player_id,
+            tuple(map(lambda t: t.value, player_nbt["Pos"])),
+            tuple(map(lambda t: t.value, player_nbt["Rotation"])),
+        )
```
Comment (gentlegiantJGC) [inline amulet/level/formats/anvil_world/format.py:435]: This would be better structured
```py
if player_id == LOCAL_PLAYER:
    if "Player" in self.root_tag["Data"]:
        # load local player
    else:
        raise KeyError("Local player doesn't exist")
else:
    # load other player

File: amulet/level/formats/leveldb_world/format.py [hunk 6]
```diff
@@ -310,3 +311,37 @@ def _get_raw_chunk_data(
...
         return self._level_manager.get_chunk_data(cx, cz, dimension)
+
+    def get_players(self) -> Generator[str, None, None]:
+        """
+        Returns a generator of all player ids that are present in the level
+        """
+        yield from (
+            pid[7:].decode("utf-8")
+            for pid, _ in self._level_manager._db.iterate(b"player_", b"player_\xFF")
+        )
```
Comment (gentlegiantJGC) [inline amulet/level/formats/leveldb_world/format.py:322]:
```py
for pid, _ in self._level_manager._db.iterate(b"player_", b"player_\xFF"):
    yield pid[7:].decode("utf-8")
```

File: amulet/level/formats/leveldb_world/format.py [hunk 7]
```diff
@@ -310,3 +311,37 @@ def _get_raw_chunk_data(
...
+        """
+        Gets the :class:`Player` object that belongs to the specified player id
+
+        If no parameter is supplied, the data of the local player will be returned
+
+        :param player_id: The desired player id
+        :return: A Player instance
+        """
+        if player_id == LOCAL_PLAYER:
+            key = player_id
```
Comment (gentlegiantJGC) [inline amulet/level/formats/leveldb_world/format.py:334]: This should be encoded to bytes as well

File: tests/test_world.py [hunk 8]
```diff
@@ -363,6 +360,45 @@ def test_save(self):
...
+            self.assertEquals(1, len(players))
+
+        def test_get_player(self):
+            if self.world.level_wrapper.platform not in (
+                "java",
+                "bedrock",
+            ):  # Only java/bedrock platforms currently support players
+                with self.assertRaises(NotImplementedError):
+                    self.world.get_players()
+                return
```
Comment (gentlegiantJGC) [inline tests/test_world.py:381]: You probably want to check if level_wrapper it is an instance of the supported classes because the platform is used to determine the format of the data. Structures such as schematic will be either Java or Bedrock based on how they were encoded.

File: amulet/api/player/player_manager.py [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+    @property
+    def level(self) -> api_level.BaseLevel:
+        """The level that this player manager is associated with."""
+        return self._level()
+
+    def get_players(self) -> Generator[str, None, None]:
+        """
+        Returns a generator of all player ids that are present in the level
+        """
+        yield from self.level.level_wrapper.get_players()
```
Comment (gentlegiantJGC) [inline amulet/api/player/player_manager.py:71]: I have just spotted that this is side-stepping the loaded and cached versions which might cause some issues.
Looking into it I will probably need to do a little bit of rewriting of the chunk manager class so that these two can share some of the logic.

For future reference this affects has_chunk and all_chunk_coords,

**Comment by gentlegiantJGC** [comment]
This is looking pretty good. I will look into sorting out the temporary system at some point.

File: amulet/api/player/player_manager.py [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+
+from amulet.api.history import Changeable
+from amulet.api.history.data_types import EntryKeyType, EntryType
+from amulet.api.history.history_manager import DatabaseHistoryManager
+from amulet.api import level as api_level
+
+LOCAL_PLAYER = "~local_player"
+
+
+class Player(Changeable):
```
Comment (gentlegiantJGC) [inline amulet/api/player/player_manager.py:15]: The player class could also do with a dimension property.

**Comment by gentlegiantJGC** [comment]
Some of these are resolved by my pull request.

**Comment by gentlegiantJGC** [comment]
I think this is almost ready to be merged.
The last few things I want are
1) The all_player_ids should also yield the local player id
2) The player class should have the dimension exposed so we can work out which player the user is in
3) I think the `uuid` property should be renamed to just `id`. On Java this is an actual uuid4 uuid but the same property is used for Bedrock which uses some other system plus potentially some other stuff as shown in my earlier message. I think `id` will show that it is unique while also not implying it is always a uuid.

**Comment by gentlegiantJGC** [comment]
Do you want to make those changes or would you like me to?

**Comment by Podshot** [comment]
I can do the rest of those changes later today

File: amulet/api/player/player_manager.py [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+        rotation: Tuple[float, float],
+        dimension: str,
+    ):
+        """
+        Creates a new instance of :class:`Player` with the given UUID, position, and rotation
+
+        :param player_id: The ID of the player
+        :param position: The position of the player in world coordinates
+        :param rotation: The rotation of the player
+        """
```
Comment (gentlegiantJGC) [inline amulet/api/player/player_manager.py:30]: Dimension is missing from the docstring

**Comment by gentlegiantJGC** [comment]
The only issue I can see is that the way you are looking up the dimension ids might cause some issues.
It might work in Java out of coincidence but we should set up some proper way of getting the dimension from the private id.

Currently our Java system stores the level data under the folder path name `""` for the overworld `"DIM-1"` for the nether and `"DIM1"` for the end. The keys of this dictionary is what is returned in the `dimensions` property that you are using so if the dimension numerical id matches up with the index of the dimension it is a coincidence.

There is a private method to convert the public facing dimension name (which should also be unique) back to the private name but there currently isn't anything to go the other way or to the private numerical id.

Bedrock is probably a little easier because the level data is stored under the numerical value rather than a weird folder name.

I think for simplicity we should store all the level data under the numerical index and have a public way of switching between that and the public name.
Each class can then handle the folder stuff how it likes

**Comment by gentlegiantJGC** [comment]
Here is my research so far

1.12 forge
	level.dat["forge"]["DimensionData"]["UsedIDs"] = [I;0, 1, 7]
		an int array of the biomes that are used.
		This relates to the DIM{X} format.
		Not sure what the 0 and 1 are for.
	
	boss data
	level.dat["Data"]["DimensionData"] = {"7": {"CustomSeed": 0L}}
	
	level.dat["Data"]["Player"]["Dimension"] = 7
		-1 = the nether
		0 = overworld
		1 = the end
?
Java 1.16 vanilla
	level.dat["Data"]["WorldGenSettings"]["dimensions"] = {"namespace:basename": {...}, "minecraft:overworld": {...}, ...}
		keys are string dimension ids.
                It looks like the dimension data is stored in `./dimensions/namespace/basename`
	player "Dimension" is a string
Java 1.17 vanilla
	adds customisable min and max height values in the generator settings

Bedrock has always been numerical. 0, 1, 2 for overworld, nether and end respectively. It doesn't yet have custom dimensions.

TLDR:
Java
before 1.16 the dimension was stored as an int. If it was 0 it was the dimension in the root directory and if it wasn't 0 it matched up with the `DIM{X}` format.
Java 1.16 switched to namespaced string dimension ids and added the ability to add new dimensions without modded.

Our system needs to handle everything.

**Comment by gentlegiantJGC** [comment]
I have decided to put off the dimension changes because it is going to require quite a bit of rewriting.
I have put together a temporary solution which I think should work for now.
